### ISSUE TITLE
How the modifications in the UI are passed to the physics simulation ?

### ISSUE THREAD

**Issue by joaopfg** [deepmind/mujoco#569]
Hi, 

I'm having a hard time trying to understand how the modifications in the UI are passed to the physics simulation in the boilerplate `simulate.cc`. I want to add new buttons to the UI to make my robot follow hard coded plans using the joint position actuators. 

So far, I understand that the physics loop runs in a background thread and modify the state through the function `Simulate::applyforceperturbations`, which calls `mjv_applyPerturbForce` to modify the `mjData` object using an `mjvPerturb` object. 

I also understand that the render loop runs in the main thread and calls `mjui_event` and `mjui_update` when I modify joint positions.

But I can't understand where the connection between the render loop and the physics loop happens. More specifically, what is the mechanism of the render loop that fills the `mjvPerturb` object so that the physics loop can modify the state of the simulation accordingly ?

A detailed high level detailed explanation would be much appreciated.

Thanks in advance

**Comment by joaopfg** [comment]
@yuvaltassa I think my problem for now is mainly with question 1 (although I also don't know the answer for questions 2 and 3). I want to modify the boilerplate `simulate.cc` to add a new button to the right panel which, when clicked, change the inputs of the positions actuators according a hard coded set of joint positions, wait until the robot goes to the final desired position (in joint space) and print to stdout the final position of the robot end-effector in cartesian space.

**Comment by yuvaltassa** [comment]
Here is a commit where a button was added to `simulate` 80b4ffdd6acedad2bf0b2def3056de13ea8cc803

**Comment by joaopfg** [comment]
@yuvaltassa Thanks for the starting point. I'm trying something similar for my button but I have a doubt. In the case of the commit you showed, the action triggered by the button is a passive one (dump a screenshot to the filesystem). However, the action I need to trigger is an active one (render a robot movement when I modify joints), which I think is more difficult. I'm trying to put prints for some hours in the simulation code to grasp what is happening but it's still obscure to me. 

I will try to summarize what I think I understand and what I don't understand.

-----------------------------------------------------------------------------------------------------------

- What I think I understand:

When I modify a joint in the boilerplate UI, the function `Simulate::uiEvent `is called, which ends up calling `mjui_event` (`mjuiItem* it = mjui_event(&sim->ui1, state, &sim->con);`). This seems to be what receives the control input from the UI and transmits to the physics loop.

On the other hand, in the physics loop (running in the background thread and called by the function `PhysicsLoop` in `main.cc`), the lines that concretely update the physics state with the new joint positions are:

```
mju_zero(d->xfrc_applied, 6*m->nbody);
sim.applyforceperturbations(); 
```

which ends up calling `mjv_applyPerturbForce` (`mjv_applyPerturbForce(this->m, this->d, &this->pert);`).

-----------------------------------------------------------------------------------------------------------

- What I don't understand:

**What is the relation between `mjui_event` and `mjv_applyPerturbForce` ?**

**After I give a control input to the position actuator in the GUI, how `mjv_applyPerturbForce` ends up receiving the right perturbation from `mjui_event` in order to update the physics state ?**

**How the modification in the physics state caused by `mjv_applyPerturbForce` ends up being rendered by the UI ?**

**Comment by joaopfg** [comment]
@yuvaltassa At a higher level, I guess my question is about the connection between the physics engine and the abstract UI. If you can provide a higher level explanation about what happens from the moment I modify a control input through the slide button in the right panel until the moment the scene is rendered in the central screen, I think would be a big help.

**Comment by yuvaltassa** [comment]
Let's start with some basic clarifications. Watch [this video](https://mujoco.readthedocs.io/en/latest/programming.html#sasimulate).

1. In seconds 8-17 you are seeing user perturbations (first in paused mode, the in play mode). This is what `mjv_applyPerturb{Pose, Force}` are used for, respectively. It is unrelated to the joint/actuator sliders.
2. Joint/Control sliders, shown in seconds 30-50 do something else, they directly modify `mjData`. This happens in [this line](https://github.com/deepmind/mujoco/blob/0b45129ca4af6f294963c1cbe520a8c87e67fb83/simulate/simulate.cc#L811) (joints) and [this line](https://github.com/deepmind/mujoco/blob/0b45129ca4af6f294963c1cbe520a8c87e67fb83/simulate/simulate.cc#L861) (controls). When constructing the slider, we give it a pointer, directly into `mjData` and are effectively saying "the slider should write into this pointer". 
3. The "Clear all" button at the top of the Control section cannot be associated with a pointer, the action that it performs (setting `mjData.ctrl` to 0) needs to be explicitly handled in the `uiEvent` function [here](https://github.com/deepmind/mujoco/blob/0b45129ca4af6f294963c1cbe520a8c87e67fb83/simulate/simulate.cc#L1293).

So basically I don't think you need to think about the perturb functions at all. Play with the code a bit more, and see if things make more sense to you. I'll close this issue but will keep following it if you have more questions.

**Comment by joaopfg** [comment]
@yuvaltassa Thanks a lot ! I inserted a new button that can make the robot go to a hard coded position when clicked. But I'm not sure if I'm doing it in the idiom expected by MuJoCo. I'm doing it in the following way:

```
      // Go to a hardcoded position
      if (it->sectionid==SECT_NEW_BUTTON && it->itemid==0) {
          int ctrl_id1 = mj_name2id(m, mjOBJ_ACTUATOR, "joint1");
          int ctrl_id2 = mj_name2id(m, mjOBJ_ACTUATOR, "joint2");
          int ctrl_id3 = mj_name2id(m, mjOBJ_ACTUATOR, "joint3");
          int ctrl_id4 = mj_name2id(m, mjOBJ_ACTUATOR, "joint4");
          int ctrl_id5 = mj_name2id(m, mjOBJ_ACTUATOR, "joint5");
          int ctrl_id6 = mj_name2id(m, mjOBJ_ACTUATOR, "joint6");

          if( ctrl_id1 >= 0 && ctrl_id2 >= 0 && ctrl_id3 >= 0 && ctrl_id4 >= 0 && ctrl_id5 >= 0 && ctrl_id6 >= 0){
              d->ctrl[ctrl_id1] = 0.5;
              d->ctrl[ctrl_id2] = 0.5;
              d->ctrl[ctrl_id3] = 0.5;
              d->ctrl[ctrl_id4] = 0.5;
              d->ctrl[ctrl_id5] = 0.5;
              d->ctrl[ctrl_id6] = 0.5;
          }

          mjui_update(SECT_NEW_BUTTON, -1, &sim->ui1, &sim->uistate, &sim->con);
      }
```

Is it the right way of modifying `d->ctrl` ? Also, how can I make it go through a path in the joint space ? I guess now my problem is the question 3 that you mentioned in your first message :)

**Comment by yuvaltassa** [comment]
What you are doing seems perfectly fine. Yes, you are setting `ctrl` correctly.

The answer to this question is a **control callback**. This is explained [here](https://mujoco.readthedocs.io/en/latest/programming.html#simulation-loop) and you can find an example [here](https://github.com/deepmind/mujoco/blob/601963beb2089f88d64682079b25e4a12907eb11/test/engine/engine_forward_test.cc#L370). You need to write [a function](https://github.com/deepmind/mujoco/blob/601963beb2089f88d64682079b25e4a12907eb11/test/engine/engine_forward_test.cc#L348) that accepts `const mjModel*` and `mjData*`, and writes whatever it wants into `ctrl`.

If you go this route then you don't need a button since it already exists. Under `Physics -> Disable Flags` you can switch Actuation on and off which will enable/disable your controller (and the actuators themselves). You can set this flag in XML as well, if you want it disabled at load time.

**Comment by joaopfg** [comment]
@yuvaltassa So, if I understood, if I want to set simple controls, like setting hard coded inputs to the position actuators, I can just set `d->ctrl` in the `uiEvent` callback directly. But if want a controller that depends on intermediate results of the simulation, then it's mandatory to set a control callback. Is it right ?

I have two doubts:

1) If I want to go through a hard coded path in the joint space, I would have to sleep a bit between each two consecutive points in the path, right ? Is there some custom cross platform sleep function in MuJoCo for these kind of cases ? I saw that in C, the sleep function for windows is taken from `windows.h` while for linux it's taken from `unistd.h`.

2) f I want to do it with the control callback, I would have to call `mjui_update` inside the callback, right ? But the callback only takes as arguments `mjModel` and `mjData`. How to proceed to have access to the other parameters `mjui_update` needs inside the control callback ?

**Comment by yuvaltassa** [comment]
0. Not mandatory. If you hack in simulate you can just read values from the global `mjData`. But that's not recommended.
1. No need for that. Just look at `mjData.time`. `simulate` will take care of the synchronisation.
2. Do **what** with the control callback? Not sure what you are asking but it's easy to pass information to the callback via `mjData.userdata`, if you need to do that.

**Comment by joaopfg** [comment]
@yuvaltassa I'm trying to execute the hard coded path after clicking the added button by playing with `mjData.time` but the robot gets frozen during the time it should execute the path and only executes the last step of the path. Is it because `mjData.time` advances too fast compared to the render loop ? How to proceed ? Should I make `mjData.time` advance slower while I'm executing the hard coded path and then return the advancement rate to what it was before the hard coded path once the path is completed ? How to do that ? Here is my code:

```
      // Go to a hardcoded position
      if (it->sectionid==SECT_NEW_BUTTON && it->itemid==0) {
          mjtNum path[10][6] = {
                  {0.5, 0.5, 0.5, 0.5, 0.5, 0.5},
                  {1, 1, 1, 1, 1, 1},
                  {0.5, 0.5, 0.5, 0.5, 0.5, 0.5},
                  {1, 1, 1, 1, 1, 1},
                  {0.5, 0.5, 0.5, 0.5, 0.5, 0.5},
                  {1, 1, 1, 1, 1, 1},
                  {0.5, 0.5, 0.5, 0.5, 0.5, 0.5},
                  {1, 1, 1, 1, 1, 1},
                  {0.5, 0.5, 0.5, 0.5, 0.5, 0.5},
                  {1, 1, 1, 1, 1, 1}
          };

          bool finished = false;
          mjtNum start_time = d->time, delta = 1.0;
          int iter = 0;

          while(d->time < start_time + 10*delta){
              if(d->time >= start_time + iter*delta){
                  int ctrl_id1 = mj_name2id(m, mjOBJ_ACTUATOR, "joint1");
                  int ctrl_id2 = mj_name2id(m, mjOBJ_ACTUATOR, "joint2");
                  int ctrl_id3 = mj_name2id(m, mjOBJ_ACTUATOR, "joint3");
                  int ctrl_id4 = mj_name2id(m, mjOBJ_ACTUATOR, "joint4");
                  int ctrl_id5 = mj_name2id(m, mjOBJ_ACTUATOR, "joint5");
                  int ctrl_id6 = mj_name2id(m, mjOBJ_ACTUATOR, "joint6");

                  if( ctrl_id1 >= 0 && ctrl_id2 >= 0 && ctrl_id3 >= 0 && ctrl_id4 >= 0 && ctrl_id5 >= 0 && ctrl_id6 >= 0){
                      d->ctrl[ctrl_id1] = path[iter][0];
                      d->ctrl[ctrl_id2] = path[iter][1];
                      d->ctrl[ctrl_id3] = path[iter][2];
                      d->ctrl[ctrl_id4] = path[iter][3];
                      d->ctrl[ctrl_id5] = path[iter][4];
                      d->ctrl[ctrl_id6] = path[iter][5];
                  }

                  mjui_update(SECT_NEW_BUTTON, -1, &sim->ui1, &sim->uistate, &sim->con);
                  ++iter;

                  if(iter == 10) break;
              }

              mj_step(m, d);
          }
      }
```

**Comment by yuvaltassa** [comment]
This is inside a control callback?
Where is this code running?

**Comment by joaopfg** [comment]
@yuvaltassa This piece of code is in the `uiEvent` callback of the `simulate.cc`. It is a slight modification of my previous piece of code (here: https://github.com/deepmind/mujoco/issues/569#issuecomment-1310281607) in order to execute a complete path instead of just one move.

**Comment by yuvaltassa** [comment]
Your code should only look at data.time and write into ctrl. You shouldn't be stepping, that already happens elsewhere.

**Comment by joaopfg** [comment]
@yuvaltassa That is what I was trying before. But `d->time` stays constant for some reason and the loop becomes infinity if I remove the `mj_step` line.

**Comment by yuvaltassa** [comment]
This is because it only runs when you click the button. Use a control callback as I explained earlier.

**Comment by joaopfg** [comment]
@yuvaltassa I see that `mjData.userdata` has type `mjtNum*`. How to proceed if I want to pass data that doesn't fit in `mjtNum` type in the control callback ? 

For example, `mjui_update` is called as `mjui_update(SECT_NEW_BUTTON, -1, &sim->ui1, &sim->uistate, &sim->con);`. How can I call it inside the control callback ?

**Comment by joaopfg** [comment]
@yuvaltassa I got something close to work. But there is still a problem with the memory allocation. The robot executes the hard coded path if I click on the button for the first time but don't execute it for the next times.

Inside the `uiEvent` callaback I have the following piece of code:

```
            if (it->sectionid==SECT_NEW_BUTTON && it->itemid==0) {
                mjMARKSTACK
                mjtNum* userdata = mj_stackAlloc(d, 2);
                userdata[0] = d->time;
                userdata[1] = 0;
                d->userdata = userdata;
                mjcb_control = control_callback;

                while(d->userdata[1] < 10){
                    ++d->userdata[1];
                    mjui_update(SECT_NEW_BUTTON, -1, &sim->ui1, &sim->uistate, &sim->con);
                }

                mjFREESTACK
            }
```

And my control callback is:

```
void control_callback(const mjModel* m, mjData *d) {
    mjtNum path[10][6] = {
            {0.5, 0.5, 0.5, 0.5, 0.5, 0.5},
            {1, 1, 1, 1, 1, 1},
            {0.5, 0.5, 0.5, 0.5, 0.5, 0.5},
            {1, 1, 1, 1, 1, 1},
            {0.5, 0.5, 0.5, 0.5, 0.5, 0.5},
            {1, 1, 1, 1, 1, 1},
            {0.5, 0.5, 0.5, 0.5, 0.5, 0.5},
            {1, 1, 1, 1, 1, 1},
            {0.5, 0.5, 0.5, 0.5, 0.5, 0.5},
            {1, 1, 1, 1, 1, 1}
    };

    mjtNum start_time = d->userdata[0], delta = 1.0;
    int iter = (int)d->userdata[1];

    if(d->time >= start_time + iter*delta && iter < 10){
        int ctrl_id1 = mj_name2id(m, mjOBJ_ACTUATOR, "joint1");
        int ctrl_id2 = mj_name2id(m, mjOBJ_ACTUATOR, "joint2");
        int ctrl_id3 = mj_name2id(m, mjOBJ_ACTUATOR, "joint3");
        int ctrl_id4 = mj_name2id(m, mjOBJ_ACTUATOR, "joint4");
        int ctrl_id5 = mj_name2id(m, mjOBJ_ACTUATOR, "joint5");
        int ctrl_id6 = mj_name2id(m, mjOBJ_ACTUATOR, "joint6i");

        if( ctrl_id1 >= 0 && ctrl_id2 >= 0 && ctrl_id3 >= 0 && ctrl_id4 >= 0 && ctrl_id5 >= 0 && ctrl_id6 >= 0){
            d->ctrl[ctrl_id1] = path[iter][0];
            d->ctrl[ctrl_id2] = path[iter][1];
            d->ctrl[ctrl_id3] = path[iter][2];
            d->ctrl[ctrl_id4] = path[iter][3];
            d->ctrl[ctrl_id5] = path[iter][4];
            d->ctrl[ctrl_id6] = path[iter][5];
        }
    }
}
```

Any idea about what I'm doing wrong ?

**Comment by joaopfg** [comment]
@yuvaltassa You can ignore my two last messages, I think I got it now. It behaves like I want. Could you indicate me if there is some problem with my approach, please ?

My control callback is:

```
void control_callback(const mjModel* m, mjData *d) {
    mjtNum path[10][6] = {
            {0.5, 0.5, 0.5, 0.5, 0.5, 0.5},
            {1, 1, 1, 1, 1, 1},
            {0.5, 0.5, 0.5, 0.5, 0.5, 0.5},
            {1, 1, 1, 1, 1, 1},
            {0.5, 0.5, 0.5, 0.5, 0.5, 0.5},
            {1, 1, 1, 1, 1, 1},
            {0.5, 0.5, 0.5, 0.5, 0.5, 0.5},
            {1, 1, 1, 1, 1, 1},
            {0.5, 0.5, 0.5, 0.5, 0.5, 0.5},
            {1, 1, 1, 1, 1, 1}
    };

    mjtNum delta = 1.0;

    if(d->time >= d->userdata[0] + delta * d->userdata[1] && d->userdata[1] < 10){
        int ctrl_id1 = mj_name2id(m, mjOBJ_ACTUATOR, "joint1");
        int ctrl_id2 = mj_name2id(m, mjOBJ_ACTUATOR, "joint2");
        int ctrl_id3 = mj_name2id(m, mjOBJ_ACTUATOR, "joint3");
        int ctrl_id4 = mj_name2id(m, mjOBJ_ACTUATOR, "joint4");
        int ctrl_id5 = mj_name2id(m, mjOBJ_ACTUATOR, "joint5");
        int ctrl_id6 = mj_name2id(m, mjOBJ_ACTUATOR, "joint6");

        if( ctrl_id1 >= 0 && ctrl_id2 >= 0 && ctrl_id3 >= 0 && ctrl_id4 >= 0 && ctrl_id5 >= 0 && ctrl_id6 >= 0){
            d->ctrl[ctrl_id1] = path[(size_t)d->userdata[1]][0];
            d->ctrl[ctrl_id2] = path[(size_t)d->userdata[1]][1];
            d->ctrl[ctrl_id3] = path[(size_t)d->userdata[1]][2];
            d->ctrl[ctrl_id4] = path[(size_t)d->userdata[1]][3];
            d->ctrl[ctrl_id5] = path[(size_t)d->userdata[1]][4];
            d->ctrl[ctrl_id6] = path[(size_t)d->userdata[1]][5];
        }

        ++d->userdata[1];
    }
}
```

And the piece of code in the `uiEvent` callback is:

```
if (it->sectionid==SECT_NEW_BUTTON && it->itemid==0) {
    d->userdata[0] = d->time;
    d->userdata[1] = 0;
    mjcb_control = control_callback;
    mjui_update(SECT_NEW_BUTTON, -1, &sim->ui1, &sim->uistate, &sim->con);
}
```

**Comment by yuvaltassa** [comment]
You, that makes sense to me :slightly_smiling_face:
